### PR TITLE
node/processor: performance improvements

### PR DESCRIPTION
**Pull request by tbjump**
Some performance improvements that don't change the protocol semantics. 

Benchmark on my machine (64 cores, 19 guardians, 1000 messages):
Before: 100 msg/s
After: 133 msg/s

### CODE REVIEW

**Comment by tbjump** [comment]
Current dependencies on/for this PR:
* main
  * **PR #3097** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3097" target="_blank">Graphite</a> 
    * **PR #3096** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3096" target="_blank">Graphite</a> 
      * **PR #3024** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3024" target="_blank">Graphite</a> 
        * **PR #3067** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3067" target="_blank">Graphite</a> 
          * **PR #3134** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3134" target="_blank">Graphite</a> 
            * **PR #3127** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3127" target="_blank">Graphite</a> 
              * **PR #3128** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3128" target="_blank">Graphite</a> 
                * **PR #3129** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3129" target="_blank">Graphite</a> 
                  * **PR #3135** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3135" target="_blank">Graphite</a> 
                    * **PR #3187** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3187" target="_blank">Graphite</a> 
                    * **PR #3139** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3139" target="_blank">Graphite</a> 
                      * **PR #3168** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3168" target="_blank">Graphite</a> 
                        * **PR #3186** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3186" target="_blank">Graphite</a> 
                          * **PR #3197** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3197" target="_blank">Graphite</a>  👈
                      * **PR #3140** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3140" target="_blank">Graphite</a> 
                    * **PR #3137** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3137" target="_blank">Graphite</a> 
                  * **PR #3130** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3130" target="_blank">Graphite</a> 
                    * **PR #3136** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3136" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3197?utm_source=stack-comment).

**Comment by tbjump** [comment]
Current dependencies on/for this PR:
* main
  * **PR #3097** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3097" target="_blank">Graphite</a> 
    * **PR #3096** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3096" target="_blank">Graphite</a> 
      * **PR #3024** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3024" target="_blank">Graphite</a> 
        * **PR #3067** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3067" target="_blank">Graphite</a> 
          * **PR #3134** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3134" target="_blank">Graphite</a> 
            * **PR #3127** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3127" target="_blank">Graphite</a> 
              * **PR #3128** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3128" target="_blank">Graphite</a> 
                * **PR #3129** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3129" target="_blank">Graphite</a> 
                  * **PR #3135** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3135" target="_blank">Graphite</a> 
                    * **PR #3187** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3187" target="_blank">Graphite</a> 
                    * **PR #3139** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3139" target="_blank">Graphite</a> 
                      * **PR #3168** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3168" target="_blank">Graphite</a> 
                        * **PR #3186** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3186" target="_blank">Graphite</a> 
                          * **PR #3197** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3197" target="_blank">Graphite</a>  👈
                      * **PR #3140** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3140" target="_blank">Graphite</a> 
                    * **PR #3137** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3137" target="_blank">Graphite</a> 
                  * **PR #3130** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3130" target="_blank">Graphite</a> 
                    * **PR #3136** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3136" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3197?utm_source=stack-comment).

**Comment by tbjump** [comment]
Current dependencies on/for this PR:
* main
  * **PR #3097** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3097" target="_blank">Graphite</a> 
    * **PR #3096** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3096" target="_blank">Graphite</a> 
      * **PR #3024** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3024" target="_blank">Graphite</a> 
        * **PR #3067** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3067" target="_blank">Graphite</a> 
          * **PR #3134** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3134" target="_blank">Graphite</a> 
            * **PR #3127** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3127" target="_blank">Graphite</a> 
              * **PR #3128** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3128" target="_blank">Graphite</a> 
                * **PR #3129** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3129" target="_blank">Graphite</a> 
                  * **PR #3135** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3135" target="_blank">Graphite</a> 
                    * **PR #3187** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3187" target="_blank">Graphite</a> 
                    * **PR #3139** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3139" target="_blank">Graphite</a> 
                      * **PR #3168** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3168" target="_blank">Graphite</a> 
                        * **PR #3186** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3186" target="_blank">Graphite</a> 
                          * **PR #3197** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3197" target="_blank">Graphite</a>  👈
                      * **PR #3140** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3140" target="_blank">Graphite</a> 
                    * **PR #3137** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3137" target="_blank">Graphite</a> 
                  * **PR #3130** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3130" target="_blank">Graphite</a> 
                    * **PR #3136** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3136" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3197?utm_source=stack-comment).

**Comment by tbjump** [comment]
Current dependencies on/for this PR:
* main
  * **PR #3097** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3097" target="_blank">Graphite</a> 
    * **PR #3096** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3096" target="_blank">Graphite</a> 
      * **PR #3024** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3024" target="_blank">Graphite</a> 
        * **PR #3067** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3067" target="_blank">Graphite</a> 
          * **PR #3134** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3134" target="_blank">Graphite</a> 
            * **PR #3127** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3127" target="_blank">Graphite</a> 
              * **PR #3128** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3128" target="_blank">Graphite</a> 
                * **PR #3129** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3129" target="_blank">Graphite</a> 
                  * **PR #3135** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3135" target="_blank">Graphite</a> 
                    * **PR #3187** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3187" target="_blank">Graphite</a> 
                    * **PR #3139** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3139" target="_blank">Graphite</a> 
                      * **PR #3168** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3168" target="_blank">Graphite</a> 
                        * **PR #3186** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3186" target="_blank">Graphite</a> 
                          * **PR #3197** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3197" target="_blank">Graphite</a>  👈
                      * **PR #3140** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3140" target="_blank">Graphite</a> 
                    * **PR #3137** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3137" target="_blank">Graphite</a> 
                  * **PR #3130** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3130" target="_blank">Graphite</a> 
                    * **PR #3136** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3136" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3197?utm_source=stack-comment).

**Review by bruce-riley — COMMENTED** [review]
Has conflicts.

File: node/pkg/db/db.go [hunk 1]
```diff
@@ -128,6 +128,17 @@ func (d *Database) StoreSignedVAA(v *vaa.VAA) error {
 	return nil
 }
 
+func (d *Database) HasVAA(id VAAID) bool {
```
Comment (bruce-riley) [inline node/pkg/db/db.go:131]: This should return an error if it is anything other than `ErrKeyNotFound`.

File: node/pkg/processor/cleanup.go [hunk 2]
```diff
@@ -79,7 +79,7 @@ func (p *Processor) handleCleanup(ctx context.Context) {
 			// This occurs when we observed a message after the cluster has already reached
 			// consensus on it, causing us to never achieve quorum.
 			if ourVaa, ok := s.ourObservation.(*VAA); ok {
-				if _, err := p.getSignedVAA(*db.VaaIDFromVAA(&ourVaa.VAA)); err == nil {
+				if p.haveSignedVAA(*db.VaaIDFromVAA(&ourVaa.VAA)) {
```
Comment (bruce-riley) [inline node/pkg/processor/cleanup.go:82]: This is ignoring other errors which where logged before.

File: node/pkg/processor/observation.go [hunk 3]
```diff
@@ -232,27 +263,18 @@ func (p *Processor) handleInboundSignedVAAWithQuorum(ctx context.Context, m *gos
...
-		p.logger.Debug("ignored SignedVAAWithQuorum message for VAA we already stored",
-			zap.String("vaaID", string(db.VaaIDFromVAA(v).Bytes())),
-		)
-		return
-	} else if err != db.ErrVAANotFound {
-		p.logger.Error("failed to look up VAA in database",
-			zap.String("vaaID", string(db.VaaIDFromVAA(v).Bytes())),
-			zap.Error(err),
-		)
+	if p.haveSignedVAA(*db.VaaIDFromVAA(v)) {
```
Comment (bruce-riley) [inline node/pkg/processor/observation.go:266]: Again, lost the check for errors.

File: node/pkg/processor/observation.go [hunk 4]
```diff
@@ -160,55 +195,51 @@ func (p *Processor) handleObservation(ctx context.Context, obs *node_common.MsgW
...
+			zap.String("digest", hash),
+			zap.String("messageId", m.MessageId),
+		)
+		return
 	}
+	// Now we *may* have quorum, depending on the guardian set in use.
+	// Let's construct the VAA and check if we actually have quorum.
 
-	if p.state.signatures[hash].ourObservation != nil {
+	sigsVaaFormat, agg := signaturesToVaaFormat(s.signatures, gs.Keys)
```
Comment (bruce-riley) [inline node/pkg/processor/observation.go:223]: I do not like this change. Please either revert it or move it to a separate PR that can be reviewed / discussed individually.

Comment (tbjump) [reply node/pkg/processor/observation.go:223]: I assume you're referring to https://github.com/wormhole-foundation/wormhole/pull/3197/commits/cedf65b4267ef3cf36f5aa88a77a131740f0b91d? 

This commit 
1. moves some code to a helper function signaturesToVaaFormat() for increased readability. No changes to the code. 
2. Adds an early return `if len(s.signatures) < quorum {return}`

Why do you not like it?

File: node/pkg/db/db.go (hunk 1 above)
Comment (tbjump) [reply node/pkg/db/db.go:131]: fixed

File: node/pkg/processor/cleanup.go (hunk 2 above)
Comment (tbjump) [reply node/pkg/processor/cleanup.go:82]: I added error logging to `haveSignedVAA()`

File: node/pkg/processor/observation.go (hunk 3 above)
Comment (tbjump) [reply node/pkg/processor/observation.go:266]: I added error logging to `haveSignedVAA()`

File: node/pkg/db/db.go (hunk 1 above)
Comment (bruce-riley) [reply node/pkg/db/db.go:131]: I don't see a fix.

File: node/pkg/processor/observation.go (hunk 4 above)
Comment (bruce-riley) [reply node/pkg/processor/observation.go:223]: A) I don't see the need to create a helper function that is only called once. Just leave the code inline like it was.
B) Since you are touching the code, we should drop `agg` all together since it is only used in debugs.
C) You are computing `quorum` twice.
D) I am not convinced that that early quorum check is the equivalent to what is down below.
E) You don't really need to move the quorum check up if you move the building of `sigs` into the check for `ouObservation` being null.

**Review by bruce-riley — CHANGES_REQUESTED** [review]

File: node/pkg/processor/observation.go (hunk 4 above)
Comment (tbjump) [reply node/pkg/processor/observation.go:223]: A) I created the helper function from because of @kcsongor 's feedback to make the code more readable and I have to agree that it improves readability. Keep your functions short and all that. 
B) I've found `agg` very useful for debugging in the past, I'd like to keep it. 
C and D) I've removed the second re-calculation of `quorum`. The quorum checks are indeed not equivalent. The first one is faster though and we can return early in most cases. This is the performance improvement. Constructing the VAA actually takes a lot of time in the benchmarks and skipping that in most cases is where most of the performance improvement in this PR is coming from. I added additional comments to explain this. 
E) I've moved the call to `signaturesToVaaFormat` into the `if s.ourObservation != nil` branch. Good call!

**Comment by tbjump** [comment]
changing to draft since this needs to be re-based and discussed.

**Comment by tbjump** [comment]
resolved merge conflicts

File: node/pkg/processor/observation.go [hunk 5]
```diff
@@ -161,63 +195,63 @@ func (p *Processor) handleObservation(ctx context.Context, obs *node_common.MsgW
...
-		}
+	quorum := vaa.CalculateQuorum(len(gs.Keys))
 
-		agg[i] = ok
+	// Check if we have more signatures than required for quorum.
+	// s.signatures may contain signatures from multiple guardian sets during guardian set updates
+	// Hence, if len(s.signatures) < quorum, then there is definitely no quorum and we can return early to save additional computation,
+	// but if len(s.signatures) >= quorum, there is not necessarily quorum for the active guardian set.
+	// We will later check for quorum again after assembling the VAA for a particular guardian set.
+	if len(s.signatures) < quorum {
```
[the remark this answers is not in the document]
Comment (tbjump) [reply node/pkg/processor/observation.go]: Good catch. Fixed.

**Review by bruce-riley — APPROVED** [review]

**Review by panoel — APPROVED** [review]
